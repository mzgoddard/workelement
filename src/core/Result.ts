export class Result<T> {
  #result: T;
  constructor(result: T) {
    this.#result = result;
  }
  unwrap() {
    return this.#result;
  }
}
export class ResultError {
  #error: any;
  constructor(error: any) {
    this.#error = error;
  }
  unwrap() {
    throw this.#error;
  }
}
export function result<T>(result: T) {
  return new Result(result);
}
export function resultError(error: any) {
  return new ResultError(error);
}
