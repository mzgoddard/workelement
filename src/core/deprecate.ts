const deprecatedSet = new Set();
export function deprecate(key: string, message: (key: string) => string) {
  if (!deprecatedSet.has(key)) {
    console.warn(message(key));
    deprecatedSet.add(key);
  }
}
