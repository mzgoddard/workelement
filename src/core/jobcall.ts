import {
  MaySlug,
  Slug,
  SlugMap,
  SlugOutputStruct,
  Sluggable,
  isCacheable,
  isDependency,
  isMaySlug,
  slug,
  toSlug,
} from "./slug";
import { isPromise } from "util/types";
import { microtaskLock } from "../support/microtask-helper";
import { abbreviate } from "../support/abbreviate";
import { Result, ResultError, result, resultError } from "./Result";

export interface JobHandle<
  Input extends [] | Sluggable[] = any,
  DerivableOutput = any
> {
  (...input: Input): DerivableOutput;
}

export interface JobOptions<Input extends any[], DerivableOutput> {
  name?: string;
  slug?(...args: Input): Slug;
  deriveInput?(inputs: JobDerivableInput<Input>): Promise<Input>;
  deriveOutput?(
    output: DerivableOutput,
    slug: Slug
  ): Promise<JobOutput<DerivableOutput>>;
  slugerize?(output: DerivableOutput, slug: Slug): DerivableOutput & Sluggable;
}

type JobOptionsOf<Handle extends JobHandle> = Handle extends JobHandle<
  infer Input,
  infer DerivableOutput
>
  ? JobOptions<Input, DerivableOutput>
  : JobOptions<any, any>;

export interface JobFactoryInitializer<Handle extends JobHandle> {
  (handle: Handle, options?: JobOptionsOf<Handle>): JobFactoryOf<Handle>;
}

export interface JobFactory<
  DerivableInput extends [] | any[] = any[],
  Output = any
> {
  (...input: DerivableInput): Job<Output>;
}

export type RenderJobFactory<Factory extends JobFactory> =
  Factory extends JobFactory<infer DerivableInput, infer Output>
    ? (...input: DerivableInput) => Job<Output>
    : never;

type JobFactoryOf<Handle extends JobHandle> = Handle extends JobHandle<
  infer Input,
  infer DerivableOutput
>
  ? JobFactory<JobDerivableInput<Input>, JobOutput<DerivableOutput>>
  : never;

export interface Job<Output = any> extends MaySlug {
  __task: any;
  __input: any;
  __output: Output;
}

export type JobDerivable<Output> = Job<Output> | PromiseLike<Output>;

/** Possible inputs to a JobFactory based on what JobHandle takes. */
export type JobDerivableInput<Input extends any[]> = Input extends []
  ? []
  : Input extends [infer First, ...infer Rest]
  ? [JobDerivableInputItem<First>, ...JobDerivableInput<Rest>]
  : Input extends (infer Item)[]
  ? JobDerivableInputItem<Item>[]
  : Input;
export type JobDerivableInputItem<InputItem> = Job<InputItem> | InputItem;

/** Result after JobHandle Output is processed. */
export type JobOutput<DerivableOutput> = DerivableOutput extends JobDerivable<
  infer Output
>
  ? JobOutput<Output>
  : DerivableOutput extends {}
  ? {
      [Key in keyof DerivableOutput]: JobOutputItem<DerivableOutput[Key]>;
    } & MaySlug
  : DerivableOutput extends []
  ? []
  : DerivableOutput extends [infer First, ...infer Rest]
  ? [JobOutputItem<First>, ...JobOutputItem<Rest>] & MaySlug
  : DerivableOutput extends (infer Item)[]
  ? JobOutputItem<Item>[] & MaySlug
  : DerivableOutput extends undefined | null
  ? Slug
  : DerivableOutput & Sluggable;
type JobOutputItem<DerivableOutputItem> =
  DerivableOutputItem extends JobDerivable<infer OutputItem>
    ? OutputItem
    : DerivableOutputItem;

export type DerivedJob<J extends Job> = J extends Job<infer Output>
  ? Output
  : never;

let optionsIndex = 0;

interface Task<
  Handle extends JobHandle<Input, DerivableOutput>,
  Input extends any[] = any[],
  DerivableOutput = any
> {
  handle: Handle;
  name: string;
  slug(...args: Input): Slug;
  deriveInput(inputs: JobDerivableInput<Input>): Promise<Input>;
  deriveOutput(
    output: DerivableOutput,
    slug: Slug
  ): Promise<JobOutput<DerivableOutput>>;
  slugerize(output: DerivableOutput, slug: Slug): DerivableOutput & Sluggable;
}
export const defaultSlug =
  (initializedName: string) =>
  (...args: JobDerivableInput<any>) =>
    slug`${initializedName}(${abbreviate(
      args
        .map(toSlug)
        .map((subslug) => subslug.__slug)
        .join(",")
    )})`;
export const defaultDeriveInput = async (
  derivableInput: JobDerivableInput<any>
) => await run(Promise.all(derivableInput.map(run)));

function initTask<Handle extends JobHandle>(
  handle: Handle,
  options?: JobOptionsOf<Handle>
): Task<Handle> {
  const {
    name: maybeName,
    slug: maybeSlug,
    deriveInput: maybeDeriveInput,
    deriveOutput: maybeDeriveOutput,
    slugerize: maybeSlugerize,
  } = options ?? {};
  const taskIndex = optionsIndex++;
  const initializedName: JobOptionsOf<Handle>["name"] =
    maybeName ?? `nameless#${taskIndex}`;
  const initializedSlug: JobOptionsOf<Handle>["slug"] =
    maybeSlug ?? defaultSlug(initializedName);
  const initializedDeriveInput: JobOptionsOf<Handle>["deriveInput"] =
    maybeDeriveInput ?? defaultDeriveInput;
  const initializedDeriveOutput: JobOptionsOf<Handle>["deriveOutput"] =
    maybeDeriveOutput ??
    (async (output, slug) =>
      Object.fromEntries([
        ...(await run(
          Promise.all(
            Object.entries(await run(await run(output))).map(
              async ([key, value]) => [key, await run(await run(value))]
            )
          )
        )),
        ["toSlug", () => slug],
      ]));
  const initializedSlugerize: JobOptionsOf<Handle>["slugerize"] =
    maybeSlugerize ??
    ((output, slug) => {
      if (output && typeof output === "object" && !Array.isArray(output)) {
        return { ...output, toSlug: () => SlugOutputStruct(slug) };
      } else if (output === undefined || output === null) {
        return SlugOutputStruct(slug);
      }
      return output;
    });
  return {
    name: initializedName,
    handle,
    slug: initializedSlug,
    deriveInput: initializedDeriveInput,
    deriveOutput: initializedDeriveOutput,
    slugerize: initializedSlugerize,
  } as any;
}

export function task<Handle extends JobHandle>(
  handle: Handle,
  options?: JobOptionsOf<Handle>
): RenderJobFactory<JobFactoryOf<Handle>> {
  const task = initTask(handle, options);
  return Object.assign((...input) => ({
    __task: task,
    __input: input,
    toSlug() {
      return task.slug(...input);
    },
  }));
}

type Work<T = any> = Promise<Result<T> | ResultError>;

interface RunContextCollection {
  job: any;
  slug: Slug;
  main: RunContext | null;
  iterations: Set<RunContext>;

  readonly globalIndex: number | null;
  readonly upToDate: boolean | null;
  readonly work: Work<any> | null;
  readonly dependencies: Set<any> | null;
  readonly guards: Set<any> | null;
}

interface RunContext {
  job: any;
  slug: Slug;
  collection: RunContextCollection;
  // index: number;
  globalIndex: number;
  upToDate: boolean;
  work: Work | null;
  copyOf: RunContext | null;
  copies: Set<RunContext> | null;
  dependencies: Set<RunContext> | undefined;
  // changes: Set<any>;
  guards: Set<Guard> | undefined;
  // resources: Set<ResourceFactory> | undefined;
  parent: RunContext | null;
  // mainIteration: RunIteration | null;
  // iterations: Set<RunIteration>;
}

let activeContext = null as RunContext;
let nextContextIndex = 0;

const contextMap = new SlugMap<RunContextCollection>();

const workMap = new SlugMap<Work>();

export async function run<J>(
  job: J
): Promise<J extends Job ? DerivedJob<J> : J extends Promise<infer P> ? P : J> {
  const parentContext = activeContext;
  try {
    if (isJob(job)) {
      const task = job.__task as Task<any>;
      const slug = Slug(job);
      // console.log("begin", slug.__slug);
      let contextCollection = findOrCreateContextCollection<J>(slug, job);
      let work = contextCollection.work;
      // if (!work) {
      // let jobContext = contextMap.get(slug);
      // let work = workMap.get(slug);
      // let work = null as Work;
      if (!work || !contextCollection.upToDate) {
        // let jobContext = contextMap.get(slug);
        // if (!jobContext) {
        const jobContext: RunContext = createContext<J>(
          job,
          slug,
          contextCollection,
          parentContext
        );
        // if (isCacheable(slug)) {
        //   contextMap.set(slug, jobContext);
        // }
        // }
        // if (parentContext) {
        //   parentContext.dependencies.add(jobContext);
        // }
        await microtaskLock();
        activeContext = jobContext;
        work = (async () => {
          try {
            const input = await task.deriveInput(job.__input);
            activeContext = jobContext;
            const output = await task.handle(...input);
            activeContext = jobContext;
            return task.slugerize(await run(output), slug);
          } finally {
            if (jobContext.guards) {
              await Promise.all(
                Array.from(jobContext.guards.values(), (guardItem) => {
                  // console.log("release guard", guardItem);
                  return guardItem.release();
                })
              );
            }
          }
        })().then(result, resultError);
        jobContext.work = work;
        // work.catch(e => {console.error('error', slug.__slug, e); throw e;}).then(() => console.log("finish", slug.__slug));
        if (isCacheable(slug)) {
          // contextMap.set(slug, jobContext);
          contextCollection.main = jobContext;
        } else {
          // console.log("uncacheable", slug);
        }
      } else {
        // console.log("copy", slug.__slug);
        createRepeatContext(job, slug, contextCollection, parentContext);
      }
      const output = await work;
      // console.log("finish", slug.__slug);
      return output.unwrap();
    }
    if (isPromise(job)) {
      return (await job) as any;
    }
    if (parentContext && isMaySlug(job)) {
      const slug = Slug(job);
      if (isDependency(slug)) {
        // console.log(isJob(job), isOutput(Slug(job)), job);
        // console.log(Slug(job));
        const contextCollection = findOrCreateContextCollection(slug, job);
        const jobContext = createRepeatContext(
          job,
          slug,
          contextCollection,
          parentContext
        );
        if (isCacheable(slug)) {
          // contextMap.set(slug, jobContext);
          contextCollection.main = jobContext;
        } else {
          // console.log("uncacheable", slug);
        }
        // parentContext.dependencies.add(job);
      }
    }
    return job as any;
  } finally {
    await microtaskLock();
    activeContext = parentContext;
  }
}

export const getContext = (job: Sluggable) => contextMap.get(Slug(job));

export const getThisContext = () => activeContext;

// export const walkMapContext = <T>(context: RunContext, map: (context: RunContext, walkMap: (context: RunContext) => T[], path: string[], root: RunContext) => T) =>

export interface DependencyTree {
  slug: Slug;
  jobIndex: number;
  upToDate: boolean | undefined;
  dependencies: DependencyTree[];
}

export const getDependencyTree = (job: Job): DependencyTree => ({
  jobIndex: getContext(job)?.globalIndex ?? -1,
  slug: Slug(job),
  upToDate: getContext(job)?.upToDate ?? undefined,
  dependencies: Array.from(getContext(job)?.dependencies?.values() ?? [])
    .map((context) => context.slug)
    .map(getDependencyTree),
});

// const m = await run({});

type JobPromiseFactory<T> = JobFactory<[Promise<T>], T>;

let nextGlobalIndex = 1;

function createRepeatContext<J>(
  job: J,
  slug: Slug,
  contextCollection: RunContextCollection,
  parentContext: RunContext
) {
  const copyOf = contextCollection.main;
  const jobContext: RunContext = {
    job,
    slug,
    collection: contextCollection,
    globalIndex: nextGlobalIndex++,
    upToDate: true,
    copyOf,
    copies: null,
    work: null as Promise<any>,
    dependencies: undefined,
    guards: undefined,
    parent: parentContext,
  };
  if (copyOf) {
    if (!copyOf.copies) {
      copyOf.copies = new Set();
    }
    copyOf.copies.add(jobContext);
  }
  contextCollection.iterations.add(jobContext);
  if (parentContext) {
    parentContext.dependencies?.add(jobContext);
  }
  return jobContext;
}

function createContext<J>(
  job: J,
  slug: Slug,
  contextCollection: RunContextCollection,
  parentContext: RunContext
) {
  const jobContext: RunContext = {
    job,
    slug,
    collection: contextCollection,
    globalIndex: nextGlobalIndex++,
    upToDate: true,
    copies: null,
    copyOf: null,
    work: null as Promise<any>,
    dependencies: new Set(),
    guards: undefined,
    parent: parentContext,
  };
  contextCollection.iterations.add(jobContext);
  if (parentContext) {
    parentContext.dependencies.add(jobContext);
  }
  return jobContext;
}

function findOrCreateContextCollection<J>(slug: Slug, job: J) {
  let contextCollection = contextMap.get(slug);
  if (!contextCollection) {
    contextCollection = {
      job,
      slug,
      main: null as RunContext,
      iterations: new Set(),
      get globalIndex() {
        return this.main?.globalIndex ?? -1;
      },
      get upToDate() {
        return this.main?.upToDate ?? null;
      },
      get work() {
        return this.main?.work ?? null;
      },
      get dependencies() {
        return this.main?.dependencies ?? null;
      },
      get guards() {
        return this.main?.guards ?? null;
      },
    };
    contextMap.set(slug, contextCollection);
  }
  return contextCollection;
}

// const pp = await run(props({ data: json(source("{}")) }));

// const v = await run(promise(Promise.resolve({ id: 0 })));

// const g = await run(get(promise(Promise.resolve({ id: 0 })), "id"));

// const t = await run(then(json(source("{}")), (obj) => ({ id: 1, ...obj })));

// const clipsDefinition = await run(json(source("{}")));
function isJob(value: unknown): value is Job {
  return Boolean(value && (value as Job).__task);
}

export interface Guard {
  release(): Promise<void>;
}

export const guardScope = task(
  (job: Job) => {
    getThisContext().guards = new Set();
    return job;
  },
  {
    name: "guardScope",
    async deriveInput(inputs) {
      return inputs;
    },
  }
);

export const addGuard = (guard: Guard) => {
  let context = getThisContext();
  while (context && !context.guards) {
    context = context.parent;
  }
  if (!context) {
    context = getThisContext();
    context.guards = new Set();
  }
  context.guards.add(guard);
};

export type ResourceFactory<T> = JobFactory<[], T>;

// export const resource = <T>(handle: () => T): ResourceFactory<T> => {};

export const initResource = task(
  <T>(resource: ResourceFactory<T>, value: T) => value,
  {
    name: "initResource",
    slug(resource, value) {
      return Slug(resource);
    },
  }
) as <T>(
  resource: JobDerivableInputItem<ResourceFactory<T>>,
  value: JobDerivableInputItem<T>
) => Job<T>;

export const walkContextLeaves = function* (
  context: RunContext
): IterableIterator<RunContext> {
  for (const dep of context.dependencies) {
    if (!isCacheable(dep.slug)) {
      continue;
    }
    if (dep.dependencies?.size > 0) {
      yield* walkContextLeaves(dep);
    } else {
      yield dep;
    }
  }
};

export const addDependencies = (dependencies: (MaySlug | Slug)[]) => {
  const contexts = dependencies.map((dep) => {
    const slug = Slug(dep);
    return findOrCreateContextCollection(slug, dep);
  });
  for (const leaf of walkContextLeaves(getThisContext())) {
    for (const dep of contexts) {
      createRepeatContext(dep.job, dep.slug, dep, leaf);
    }
  }
};

export const walkContextParents = function* (
  context: RunContextCollection
): IterableIterator<RunContextCollection> {
  yield context;
  for (const iteration of context.iterations) {
    if (iteration.parent) {
      yield* walkContextParents(iteration.parent.collection);
    }
  }
};

export const announceChanges = (changes: (MaySlug | Slug)[]) => {
  for (const change of changes) {
    for (const context of walkContextParents(getContext(change))) {
      for (const iteration of context.iterations) {
        iteration.upToDate = false;
      }
      // context.upToDate = false;
      // context.main = null;
    }
  }
};
