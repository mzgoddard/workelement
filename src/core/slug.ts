import { inspect } from "util";
import { abbreviate } from "../support/abbreviate";
import { deprecate } from "./deprecate";

export const SLUG_VALUE: unique symbol = Symbol.for("workelement.slug");
export const SLUGIFY: unique symbol = Symbol.for("workelement.slugify");

type RawSlug = string;
export interface MaySlug {
  [SLUGIFY](): Slug;
}
export interface Slug extends MaySlug {
  [SLUG_VALUE]: RawSlug;
}
interface SlugCacheable extends Slug {
  isCacheable?: boolean;
}
interface SlugOutput extends Slug {
  isOutput?: boolean;
}
interface SlugDependency extends Slug {
  isDependency?: boolean;
}
export type Sluggable = MaySlug | { toString(): string } | null | undefined;
function isSlug(maybe: unknown): maybe is Slug {
  return Boolean(maybe && (maybe as Slug)[SLUG_VALUE]);
}
export function isCacheable(maybe: Slug): boolean {
  try {
    // console.log(Object.keys(maybe));
    return (maybe as SlugCacheable).isCacheable ?? true;
  } catch (e) {
    console.error(e);
    throw e;
  }
}
export function isOutput(maybe: Slug): boolean {
  return (maybe as SlugOutput).isOutput ?? false;
}
export function isDependency(maybe: Slug): boolean {
  return (maybe as SlugDependency).isDependency ?? false;
}
function inspectSlug(this: Slug & SlugCacheable & SlugDependency & SlugOutput) {
  return `[Slug ${this[SLUG_VALUE]}${this.isCacheable ? "" : " uncacheable"}${
    this.isOutput ? " output" : ""
  }${this.isDependency ? " dependency" : ""}]`;
}
class SlugBase implements Slug {
  [SLUG_VALUE]: RawSlug;
  constructor(slug: RawSlug) {
    this[SLUG_VALUE] = slug;
  }
  get __slug() {
    deprecate("SlugBase.__slug", (key) => `${key} is deprecated`);
    return this[SLUG_VALUE];
  }
  get isCacheable() {
    return true;
  }
  get isOutput() {
    return false;
  }
  get isDependency() {
    return false;
  }
  [SLUGIFY]() {
    return Slug(this[SLUG_VALUE]);
    // return this;
  }
  [inspect.custom]() {
    return inspectSlug.call(this);
  }
}
class SlugProxy implements Slug {
  target: Slug;
  constructor(target: Slug) {
    this.target = target;
  }
  get [SLUG_VALUE](): RawSlug {
    return this.target[SLUG_VALUE];
  }
  get __slug() {
    deprecate("SlugProxy.__slug", (key) => `${key} is deprecated`);
    return this[SLUG_VALUE];
  }
  get isCacheable(): boolean {
    return (this.target as SlugCacheable).isCacheable ?? true;
  }
  get isOutput(): boolean {
    return (this.target as SlugOutput).isOutput ?? false;
  }
  get isDependency(): boolean {
    return (this.target as SlugDependency).isDependency ?? false;
  }
  [SLUGIFY]() {
    return this.target[SLUGIFY]();
  }
  [inspect.custom]() {
    return inspectSlug.call(this);
  }
}
class SlugCacheableMod extends SlugProxy {
  #cacheable: boolean;
  constructor(target: Slug, cacheable: boolean) {
    super(target);
    this.#cacheable = cacheable;
  }
  get isCacheable(): boolean {
    return this.#cacheable;
  }
}
class SlugOutputMod extends SlugProxy {
  #output: boolean;
  constructor(target: Slug, output: boolean) {
    super(target);
    this.#output = output;
  }
  get isOutput(): boolean {
    return this.#output;
  }
}
class SlugDependencyMod extends SlugProxy {
  #dependency: boolean;
  constructor(target: Slug, dependency: boolean) {
    super(target);
    this.#dependency = dependency;
  }
  get isDependency(): boolean {
    return this.#dependency;
  }
}
function SlugStruct(__slug: RawSlug): Slug {
  return new SlugBase(__slug);
  return {
    [SLUG_VALUE]: __slug,
    [SLUGIFY]() {
      return Slug(__slug);
    },
    [inspect.custom]() {
      return `Slug ${this.__slug}`;
    },
  } as Slug;
}
export function SlugUncacheable(value: Slug): Slug {
  return new SlugCacheableMod(value, false);
  return { ...value, isCacheable: false } as Slug;
}
export function SlugOutputStruct(value: Slug): Slug {
  // return new SlugOutputMod(value, true);
  return { ...value, isOutput: true } as Slug;
}
export function SlugDependencyStruct(value: Slug): Slug {
  return new SlugDependencyMod(value, true);
  return { ...value, isDependency: true } as Slug;
}
export function Slug(value: RawSlug | Slug | Sluggable): Slug {
  // console.log("Slug", isSlug(value), isMaySlug(value), typeof value, value);
  if (isSlug(value)) {
    return value;
  } else if (isMaySlug(value)) {
    return value[SLUGIFY]();
  }
  return SlugStruct(String(value));
}
export function SlugArray(array: (RawSlug | Slug | Sluggable)[]): Slug {
  return SlugStruct(
    `[${array.map((value) => Slug(value)[SLUG_VALUE]).join(",")}]`
  );
}
// export function SlugObscured(name: string, variable)
function SlugLazy(handle: () => RawSlug): Slug {
  return {
    get [SLUG_VALUE]() {
      return handle();
    },
    [SLUGIFY]() {
      return SlugStruct(handle());
    },
  };
}
export function isMaySlug(maybe: unknown): maybe is MaySlug {
  // console.log("isMaySlug", maybe, (maybe as MaySlug)[SLUGIFY]);
  return Boolean(maybe && (maybe as MaySlug)[SLUGIFY]);
}
export function toSlug(sluggable: Sluggable): Slug {
  if (isMaySlug(sluggable)) {
    return sluggable[SLUGIFY]();
  }
  return SlugStruct(String(sluggable));
}
export function slug(
  strings: TemplateStringsArray,
  ...args: Sluggable[]
): Slug {
  return SlugStruct(
    strings.map((s, i) => s + Slug(args[i] ?? "")[SLUG_VALUE]).join("")
  );
}
slug.uncacheable = (strings: TemplateStringsArray, ...args: Sluggable[]) =>
  SlugUncacheable(slug(strings, ...args));
slug.dependency = (strings: TemplateStringsArray, ...args: Sluggable[]) =>
  SlugDependencyStruct(slug(strings, ...args));
export class SlugMap<V> implements Map<Slug, V> {
  map = new Map<RawSlug, V>();

  get size() {
    return this.map.size;
  }
  get [Symbol.toStringTag]() {
    return this.map[Symbol.toStringTag];
  }
  [Symbol.iterator](): IterableIterator<[Slug, V]> {
    return this.entries();
  }
  clear(): void {
    this.map.clear();
  }
  delete(key: Slug): boolean {
    return this.map.delete(key[SLUG_VALUE]);
  }
  get(key: Slug) {
    return this.map.get(key[SLUG_VALUE]);
  }
  has(key: Slug): boolean {
    return this.map.has(key[SLUG_VALUE]);
  }
  set(key: Slug, value: V): this {
    this.map.set(key[SLUG_VALUE], value);
    return this;
  }
  *entries(): IterableIterator<[Slug, V]> {
    for (const [key, value] of this.map.entries()) {
      yield [Slug(key), value];
    }
  }
  forEach(
    callbackfn: (value: V, key: Slug, map: Map<Slug, V>) => void,
    thisArg?: any
  ): void {
    this.map.forEach((value, key) => callbackfn(value, Slug(key), this));
  }
  *keys(): IterableIterator<Slug> {
    for (const key of this.map.keys()) {
      yield Slug(key);
    }
  }
  values(): IterableIterator<V> {
    return this.map.values();
  }
}

export class SlugSet<T extends MaySlug> implements Set<T> {
  map = new SlugMap<T>();

  get [Symbol.toStringTag]() {
    return this.map[Symbol.toStringTag];
  }
  [Symbol.iterator](): IterableIterator<T> {
    return this.values();
  }

  add(value: T): this;
  add(value: T): this;
  add(value: unknown): this {
    const slug = Slug(value);
    if (!this.map.has(slug)) {
      this.map.set(slug, value as T);
    }
    return this;
  }

  clear(): void;
  clear(): void;
  clear(): void {
    this.map.clear();
  }

  delete(value: T): boolean;
  delete(value: T): boolean;
  delete(value: unknown): boolean {
    const slug = Slug(value);
    if (this.map.has(slug)) {
      this.map.delete(slug);
      return true;
    }
    return false;
  }

  entries(): IterableIterator<[T, T]>;
  entries(): IterableIterator<[T, T]>;
  *entries(): IterableIterator<[T, T]> {
    for (const value of this.map.values()) {
      yield [value, value];
    }
  }

  forEach(
    callbackfn: (value: T, value2: T, set: Set<T>) => void,
    thisArg?: any
  ): void;
  forEach(
    callbackfn: (value: T, value2: T, set: Set<T>) => void,
    thisArg?: any
  ): void;
  forEach(callbackfn: unknown, thisArg?: unknown): void {
    this.map.forEach((value) =>
      (callbackfn as any).call(thisArg, value, value, this)
    );
  }

  has(value: T): boolean;
  has(value: T): boolean;
  has(value: unknown): boolean {
    const slug = Slug(value);
    return this.map.has(slug);
  }

  keys(): IterableIterator<T>;
  keys(): IterableIterator<T>;
  *keys(): IterableIterator<T> {
    yield* this.map.values();
  }

  get size(): number {
    return this.map.size;
  }

  values(): IterableIterator<T>;
  values(): IterableIterator<T>;
  *values(): IterableIterator<T> {
    yield* this.map.values();
  }
}
