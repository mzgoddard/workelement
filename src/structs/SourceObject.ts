import { MaySlug } from "../core/slug";


export interface SourceObject extends MaySlug {
  content: string;
}
