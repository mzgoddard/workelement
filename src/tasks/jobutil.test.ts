import { describe, expect, it } from "vitest";

import {
  all,
  call,
  func,
  get,
  json,
  promise,
  props,
  source,
} from "./jobutil";
import { SourceObject } from "../structs/SourceObject";
import { run } from "../core/jobcall";
import { Slug } from "../core/slug";

describe("jobutil", () => {
  it("source", async () => {
    expect(Slug(source("{}"))).to.matchSnapshot();
  });

  it("func", async () => {
    expect(Slug(func("returnhi", () => "hi"))).to.matchSnapshot();
  });

  it.each([
    [call(func("returnhi", () => "hi"))],
    [
      call(
        func("JSON.parse", (source: SourceObject) => ({
          data: JSON.parse(source.content),
        })),
        source("{}")
      ),
    ],
  ])("call", async (job) => {
    expect(job).to.matchSnapshot();
    expect(Slug(job)).to.matchSnapshot();
    await expect(run(job)).resolves.to.matchSnapshot();
    // source;
    // json;
    // promise;
    // props;
    // get;
    // // then;
  });

  it.each([[json(source("{}"))], [json(source("[]"))]])("json", async (job) => {
    expect(job).to.matchSnapshot();
    expect(Slug(job)).to.matchSnapshot();
    await expect(run(job)).resolves.to.matchSnapshot();
  });

  it.each([
    [all([])],
    [all([call(func("returnhi", () => "hi"))])],
    [all([json(source("{}"))])],
  ])("all", async (job) => {
    expect(job).to.matchSnapshot();
    expect(Slug(job)).to.matchSnapshot();
    await expect(run(job)).resolves.to.matchSnapshot();
  });
});
