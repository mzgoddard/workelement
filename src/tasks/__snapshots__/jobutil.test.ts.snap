// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`jobutil > all 1`] = `
{
  "__input": [
    [],
  ],
  "__task": {
    "deriveInput": [Function],
    "deriveOutput": [Function],
    "handle": [Function],
    "name": "all",
    "slug": [Function],
    "slugerize": [Function],
  },
  Symbol(workelement.slugify): [Function],
}
`;

exports[`jobutil > all 2`] = `
{
  "__slug": "all([])",
  Symbol(workelement.slugify): [Function],
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;

exports[`jobutil > all 3`] = `[]`;

exports[`jobutil > all 4`] = `
{
  "__input": [
    [
      {
        "__input": [
          {
            "func": [Function],
            "name": "returnhi",
            Symbol(workelement.slugify): [Function],
          },
        ],
        "__task": {
          "deriveInput": [Function],
          "deriveOutput": [Function],
          "handle": [Function],
          "name": "call",
          "slug": [Function],
          "slugerize": [Function],
        },
        Symbol(workelement.slugify): [Function],
      },
    ],
  ],
  "__task": {
    "deriveInput": [Function],
    "deriveOutput": [Function],
    "handle": [Function],
    "name": "all",
    "slug": [Function],
    "slugerize": [Function],
  },
  Symbol(workelement.slugify): [Function],
}
`;

exports[`jobutil > all 5`] = `
{
  "__slug": "all([call(func#returnhi#md5#kW2db_QbhZ_M_Pk9SSOeYw)])",
  Symbol(workelement.slugify): [Function],
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;

exports[`jobutil > all 6`] = `
[
  "hi",
]
`;

exports[`jobutil > all 7`] = `
{
  "__input": [
    [
      {
        "__input": [
          {
            "content": "{}",
            Symbol(workelement.slugify): [Function],
          },
        ],
        "__task": {
          "deriveInput": [Function],
          "deriveOutput": [Function],
          "handle": [Function],
          "name": "fromJSON",
          "slug": [Function],
          "slugerize": [Function],
        },
        Symbol(workelement.slugify): [Function],
      },
    ],
  ],
  "__task": {
    "deriveInput": [Function],
    "deriveOutput": [Function],
    "handle": [Function],
    "name": "all",
    "slug": [Function],
    "slugerize": [Function],
  },
  Symbol(workelement.slugify): [Function],
}
`;

exports[`jobutil > all 8`] = `
{
  "__slug": "all([fromJSON(source(md5#mZFLkyvTelC5g8XnyQrpOw))])",
  Symbol(workelement.slugify): [Function],
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;

exports[`jobutil > all 9`] = `
[
  {},
]
`;

exports[`jobutil > call 1`] = `
{
  "__input": [
    {
      "func": [Function],
      "name": "returnhi",
      Symbol(workelement.slugify): [Function],
    },
  ],
  "__task": {
    "deriveInput": [Function],
    "deriveOutput": [Function],
    "handle": [Function],
    "name": "call",
    "slug": [Function],
    "slugerize": [Function],
  },
  Symbol(workelement.slugify): [Function],
}
`;

exports[`jobutil > call 2`] = `
{
  "__slug": "call(func#returnhi#md5#kW2db_QbhZ_M_Pk9SSOeYw)",
  Symbol(workelement.slugify): [Function],
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;

exports[`jobutil > call 3`] = `"hi"`;

exports[`jobutil > call 4`] = `
{
  "__input": [
    {
      "func": [Function],
      "name": "JSON.parse",
      Symbol(workelement.slugify): [Function],
    },
    {
      "content": "{}",
      Symbol(workelement.slugify): [Function],
    },
  ],
  "__task": {
    "deriveInput": [Function],
    "deriveOutput": [Function],
    "handle": [Function],
    "name": "call",
    "slug": [Function],
    "slugerize": [Function],
  },
  Symbol(workelement.slugify): [Function],
}
`;

exports[`jobutil > call 5`] = `
{
  "__slug": "call(func#JSON.parse#md5#5z1TkS6-WVk73xjDQ1w3RQ,source(md5#mZFLkyvTelC5g8XnyQrpOw))",
  Symbol(workelement.slugify): [Function],
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;

exports[`jobutil > call 6`] = `{}`;

exports[`jobutil > func 1`] = `
{
  "__slug": "func#returnhi#md5#kW2db_QbhZ_M_Pk9SSOeYw",
  Symbol(workelement.slugify): [Function],
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;

exports[`jobutil > json 1`] = `
{
  "__input": [
    {
      "content": "{}",
      Symbol(workelement.slugify): [Function],
    },
  ],
  "__task": {
    "deriveInput": [Function],
    "deriveOutput": [Function],
    "handle": [Function],
    "name": "fromJSON",
    "slug": [Function],
    "slugerize": [Function],
  },
  Symbol(workelement.slugify): [Function],
}
`;

exports[`jobutil > json 2`] = `
{
  "__slug": "fromJSON(source(md5#mZFLkyvTelC5g8XnyQrpOw))",
  Symbol(workelement.slugify): [Function],
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;

exports[`jobutil > json 3`] = `{}`;

exports[`jobutil > json 4`] = `
{
  "__input": [
    {
      "content": "[]",
      Symbol(workelement.slugify): [Function],
    },
  ],
  "__task": {
    "deriveInput": [Function],
    "deriveOutput": [Function],
    "handle": [Function],
    "name": "fromJSON",
    "slug": [Function],
    "slugerize": [Function],
  },
  Symbol(workelement.slugify): [Function],
}
`;

exports[`jobutil > json 5`] = `
{
  "__slug": "fromJSON(source(md5#11FxOYiYfpMxmANj4kGJzg))",
  Symbol(workelement.slugify): [Function],
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;

exports[`jobutil > json 6`] = `{}`;

exports[`jobutil > source 1`] = `
{
  "__slug": "source(md5#mZFLkyvTelC5g8XnyQrpOw)",
  Symbol(workelement.slugify): [Function],
  Symbol(nodejs.util.inspect.custom): [Function],
}
`;
